File: database_seed/src/entity/Thread.ts
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  UpdateDateColumn,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  BaseEntity,
} from 'typeorm';
import { Keyword } from './Keyword';
import { User } from './User';

@Entity('thread')
export class Thread extends BaseEntity {
  @PrimaryGeneratedColumn('increment', { name: 'id', type: 'bigint' })
  id: string;

  @Column('varchar', { name: 'content', nullable: true, length: 5000 })
  content: string | null;

  @Column('bigint', { name: 'author_id', nullable: true })
  authorId: string | null;

  @Column('bigint', { name: 'keyword_id', nullable: true })
  keywordId: string | null;

  @Column('bigint', { name: 'parent_thread_id', nullable: true })
  parentThreadId: string | null;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @DeleteDateColumn({ name: 'deleted_at' })
  deletedAt: Date | null;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date | null;

  @ManyToOne(() => Keyword, (keyword) => keyword.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'keyword_id', referencedColumnName: 'id' }])
  keyword: Keyword;

  @ManyToOne(() => User, (user) => user.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'author_id', referencedColumnName: 'id' }])
  author: User;

  @ManyToOne(() => Thread, (thread) => thread.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'parent_thread_id', referencedColumnName: 'id' }])
  parentThread: Thread;

  @OneToMany(() => Thread, (thread) => thread.parentThread)
  threads: Thread[];
}

File: rest/src/main/java/com/waglewagle/rest/thread/entity/Thread.java
package com.waglewagle.rest.thread.entity;

import com.waglewagle.rest.keyword.entity.Keyword;
import com.waglewagle.rest.thread.data_object.dto.ThreadVO;
import com.waglewagle.rest.user.entity.User;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class Thread {

    @Id
    @GeneratedValue
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    private User author;

    @Column(length = 5000)
    private String content;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "keyword_id")
    private Keyword keyword;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_thread_id")
    private Thread parentThread;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "parentThread")
    private List<Thread> children = new ArrayList<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    private LocalDateTime deletedAt;

    @Builder
    public Thread(Long id, User author, Thread parentThread, Keyword keyword, String content) {
        this.id = id;
        this.author = author;
        this.parentThread = parentThread;
        this.keyword = keyword;
        this.content = content;
    }

    public static Thread of(ThreadVO.CreateVO createVO) {
        Thread thread = new Thread();
        thread.author = createVO.getAuthor();
        thread.parentThread = createVO.getParentThread();
        thread.content = createVO.getContent();
        thread.keyword = createVO.getKeyword();
        return thread;
    }
}
